File: schemas/content.py
# → app/schemas/content.py

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


# ── Enums ─────────────────────────────────────────────────────────────────────

class ContentType(str, Enum):
    video = "video"
    image = "image"
    audio = "audio"
    pdf = "pdf"
    text = "text"


# ── Request Schemas ───────────────────────────────────────────────────────────

class ContentCreate(BaseModel):
    """
    Used for text content created via JSON body.
    File-based content (video/image/pdf) is handled separately
    via multipart form upload in the router.
    """
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    content_type: ContentType
    text_body: Optional[str] = Field(None, description="Required when content_type is 'text'")
    plan_ids: list[int] = Field(default_factory=list, description="Gate this content to one or more plans. Empty = all subscribers.")


class ContentUpdate(BaseModel):
    """All fields optional — only what is sent gets updated."""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)
    text_body: Optional[str] = None
    thumbnail_url: Optional[str] = Field(None, max_length=1000)
    plan_ids: Optional[list[int]] = Field(None, description="Replace the plan gates. Pass empty list to remove all gates.")


# ── Response Schemas ──────────────────────────────────────────────────────────

class PlanGateSummary(BaseModel):
    """Minimal plan info shown on content so members know which plan unlocks it."""
    id: int
    name: str

    model_config = {"from_attributes": True}


class ContentResponse(BaseModel):
    """Full response — returned to the creator for their own content."""
    id: int
    hub_id: int
    plans: list[PlanGateSummary] = []
    title: str
    description: Optional[str] = None
    content_type: ContentType
    text_body: Optional[str] = None
    file_url: Optional[str]  = None
    thumbnail_url: Optional[str] = None
    is_published: bool
    is_pinned: bool
    view_count: int = 0
    like_count: int  = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ContentPublicResponse(BaseModel):
    """
    Returned to members who have access.
    Excludes is_published/is_pinned state — those are creator concerns.
    """
    id: int
    hub_id: int
    plans: list[PlanGateSummary] = []
    title: str
    description: Optional[str]     = None
    content_type: ContentType
    text_body: Optional[str]     = None
    file_url: Optional[str]     = None
    thumbnail_url: Optional[str]     = None
    is_pinned: bool
    view_count: int  = 0
    like_count: int  = 0
    is_liked: bool = False
    created_at: datetime

    model_config = {"from_attributes": True}


class LikeToggleResponse(BaseModel):
    content_id: int
    is_liked: bool
    like_count: int
    message: str

File: frontend/types/content.ts
import type { ContentType, PlanGateSummary } from "@/types/shared";

export type { ContentType };

export interface ContentResponse {
  id: number;
  hub_id: number;
  plan_id?: number | null;
  view_count: number;
  like_count: number;
  plans: PlanGateSummary[];
  title: string;
  description?: string | null;
  content_type: ContentType;
  text_body?: string | null;
  file_url?: string | null;
  thumbnail_url?: string | null;
  is_published: boolean;
  is_pinned: boolean;
  created_at: string;
  updated_at: string;
}

export interface ContentPublicResponse {
  id: number;
  hub_id: number;
  plans: PlanGateSummary[];
  title: string;
  description?: string | null;
  content_type: ContentType;
  text_body?: string | null;
  file_url?: string | null;
  thumbnail_url?: string | null;
  is_pinned: boolean;
  view_count: number; // ←
  like_count: number; // ←
  created_at: string;
  is_liked: boolean;
}

export interface ContentUpdate {
  title?: string | null;
  description?: string | null;
  text_body?: string | null;
  thumbnail_url?: string | null;
  plan_ids?: number[] | null;
}

export interface ContentCreateFields {
  title: string;
  content_type: ContentType;
  description?: string;
  text_body?: string;
  plan_id?: number;
  file?: File;
}

// ─── APPEND THESE TO YOUR EXISTING @/types/content.ts ───────────────────────

// ─── Like toggle ─────────────────────────────────────────────────────────────

export interface LikeToggleResponse {
  content_id: number;
  is_liked: boolean;
  like_count: number;
  message: string;
}

// ─── Save / bookmark toggle ─────────────────────────────────────────────────

export interface ToggleSaveResponse {
  content_id: number;
  is_saved: boolean;
  message: string;
}

// ─── Saved content list ──────────────────────────────────────────────────────

export interface SavedHubSummary {
  id: number;
  name: string;
  avatar_url?: string | null;
}

export interface SavedContentDetail {
  id: number;
  title: string;
  description?: string | null;
  content_type: "video" | "image" | "audio" | "pdf" | "text";
  thumbnail_url?: string | null;
  hub: SavedHubSummary;
  created_at: string;
}

export interface SavedContentResponse {
  id: number;
  content: SavedContentDetail;
  saved_at: string;
}

// ─── Comments ────────────────────────────────────────────────────────────────

export interface CommentAuthorSummary {
  id: string; // UUID
  username: string;
  avatar_url?: string | null;
}

export interface CommentResponse {
  id: number;
  content_id: number;
  parent_id?: number | null;
  author: CommentAuthorSummary | null;
  body: string | null;
  is_deleted: boolean;
  reply_count: number;
  created_at: string;
  updated_at: string;
}

export interface ReplyResponse {
  id: number;
  content_id: number;
  parent_id: number;
  author: CommentAuthorSummary | null;
  body: string | null;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
}

export interface CommentCreate {
  body: string;
}

export interface CommentUpdate {
  body: string;
}
